x-pack/plugins/monitoring/server/alerts/disk_usage_alert.ts

```diff
@@ -0,0 +1,403 @@
+          values: {
+            count: firingCount,
+            clusterName: cluster.clusterName,
+            action,
+          },
+        }
+      );
+
+      instance.scheduleActions('default', {
+        internalShortMessage,
+        internalFullMessage: this.isCloud ? internalShortMessage : internalFullMessage,
+        state: AlertingDefaults.ALERT_STATE.firing,
+        nodes: firingNodes
+          .map((state) => `${state.nodeName}:${state.diskUsage.toFixed(2)}`)
+          .join(','),
+        count: firingCount,
+        clusterName: cluster.clusterName,
+        action,
+        actionPlain: shortActionText,
+      });
+    } else {
+      const resolvedNodes = (alertStates as AlertDiskUsageState[])
+        .filter((state) => !state.ui.isFiring)
+        .map((state) => `${state.nodeName}:${state.diskUsage.toFixed(2)}`);
+      const resolvedCount = resolvedNodes.length;
+
+      if (resolvedCount > 0) {
+        const internalMessage = i18n.translate(
+          'xpack.monitoring.alerts.diskUsage.resolved.internalMessage',
+          {
+            defaultMessage: `Disk usage alert is resolved for {count} node(s) in cluster: {clusterName}.`,
+            values: {
+              count: resolvedCount,
+              clusterName: cluster.clusterName,
+            },
+          }
+        );
+
+        instance.scheduleActions('default', {
+          internalShortMessage: internalMessage,
+          internalFullMessage: internalMessage,
+          state: AlertingDefaults.ALERT_STATE.resolved,
+          nodes: resolvedNodes.join(','),
+          count: resolvedCount,
+          clusterName: cluster.clusterName,
+        });
+      }
+    }
+  }
+
+  private executeDeltas(
+    services: AlertServices,
+    cluster: AlertCluster,
+    newAlertStates: AlertDiskUsageState[],
+    oldAlertStates: AlertDiskUsageState[]
+  ) {
+    const deltaFiringStates = [];
+    const deltaResolvedStates = [];
+    const deltaInstanceIdPrefix: string = `.monitoring:${this.type}:${
+      cluster.clusterUuid
+    }:${Date.now()}:`;
+
+    for (const newAlertState of newAlertStates) {
+      const relatedOldState = oldAlertStates.find(
+        (oldState) =>
+          oldState.nodeId === newAlertState.nodeId &&
+          oldState.ui.isFiring !== newAlertState.ui.isFiring &&
+          oldState.ui.resolvedMS !== newAlertState.ui.resolvedMS
+      );
+      if (!relatedOldState) {
+        if (newAlertState.ui.isFiring) {
+          deltaFiringStates.push(newAlertState);
+        } else if (newAlertState.ui.resolvedMS) {
+          deltaResolvedStates.push(newAlertState);
+        }
+      }
+    }
+
+    if (deltaFiringStates.length + deltaResolvedStates.length === newAlertStates.length) {
+      /** No delta changes, so we do nothing */
+      return;
+    }
+
+    if (deltaFiringStates.length) {
+      const instance = services.alertInstanceFactory(`${deltaInstanceIdPrefix}:firing`);
```

Comment:
I like this idea _a lot_ - We can just execute actions off a unique instance id for firing and resolved